src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs

```diff
@@ -0,0 +1,171 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Threading.Tasks
+{
+    public static class TimeProviderTaskExtensions
+    {
+        private sealed class DelayState : TaskCompletionSource<bool>
+        {
+            public DelayState() : base(TaskCreationOptions.RunContinuationsAsynchronously) {}
+            public ITimer Timer { get; set; }
+            public CancellationTokenRegistration Registration { get; set; }
+        }
+
+        public static Task Delay(this TimeProvider timeProvider, TimeSpan delay, CancellationToken cancellationToken = default)
+        {
+            if (timeProvider == TimeProvider.System)
+            {
+                return Task.Delay(delay, cancellationToken);
+            }
+
+            if (timeProvider is null)
+            {
+                throw new ArgumentNullException(nameof(timeProvider));
+            }
+
+            if (delay != Timeout.InfiniteTimeSpan && delay < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(delay));
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+
+            if (delay == TimeSpan.Zero)
+            {
+                return Task.CompletedTask;
+            }
+
+            DelayState state = new();
+            state.Timer = timeProvider.CreateTimer(delayState =>
+            {
+                DelayState s = (DelayState)delayState;
+                s.TrySetResult(true);
+                s.Registration.Dispose();
+                s.Timer.Dispose();
+            }, state, Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+
+            state.Timer.Change(delay, Timeout.InfiniteTimeSpan);
+
+            state.Registration = cancellationToken.Register(delayState =>
+            {
+                DelayState s = (DelayState)delayState;
+                s.TrySetCanceled(cancellationToken);
+                s.Timer.Dispose();
+                s.Registration.Dispose();
+            }, state);
+
+            if (state.Task.IsCompleted)
```

Comment:
Similarly here, it's worth a comment about the race condition this check is here to address, that of the timer firing after we've hooked up the cancellation callback but before the registration has been stored into state.Registration; we need the subsequent check to ensure that we don't leave the registration dangling.